pennylane/math/quantum.py

```diff
@@ -1207,3 +1207,82 @@ def _compute_max_entropy(density_matrix, base):
     maximum_entropy = qml.math.log(rank) / div_base
 
     return maximum_entropy
+
+
+def trace_distance(state0, state1, check_state=False, c_dtype="complex128"):
+    r"""
+    Compute the trace distance between two quantum states.
+
+    .. math::
+        T(\rho, \sigma)=\frac12\|\rho-\sigma\|_1
+        =\frac12\text{Tr}\left(\sqrt{(\rho-\sigma)^{\dagger}(\rho-\sigma)}\right)
+
+    where :math:`\|\cdot\|_1` is the Schatten :math:`1`-norm.
+
+    The trace distance measures how close two quantum states are. In particular, it upper-bounds
+    the probability of distinguishing two quantum states.
+
+    Args:
+        state0 (tensor_like): ``(2**N, 2**N)`` or ``(batch_dim, 2**N, 2**N)`` density matrix.
+        state1 (tensor_like): ``(2**N, 2**N)`` or ``(batch_dim, 2**N, 2**N)`` density matrix.
+        check_state (bool): If True, the function will check the states' validity (shape and norm).
+        c_dtype (str): Complex floating point precision type.
+
+    Returns:
+        float: Trace distance between state0 and state1
+
+    **Examples**
+
+    The trace distance between two equal states is always zero:
+
+    >>> x = np.array([[1, 0], [0, 0]])
+    >>> qml.math.trace_distance(x, x)
+    0.0
+
+    The quantum states can also be provided as batches of density matrices:
+
+    >>> batch0 = np.array([np.eye(2) / 2, np.ones((2, 2)) / 2, np.array([[1, 0],[0, 0]])])
+    >>> batch1 = np.array([np.ones((2, 2)) / 2, np.ones((2, 2)) / 2, np.array([[1, 0],[0, 0]])])
+    >>> qml.math.trace_distance(batch0, batch1)
+    array([0.5, 0. , 0. ])
+
+    If only one of the two states represent a single element, then the trace distances are taken
+    with respect to that element:
+
+    >>> rho = np.ones((2, 2)) / 2
+    >>> sigma = rho.reshape(1, 2, 2)
+    >>> qml.math.trace_distance(rho, batch0)
+    array([0.5       , 0.        , 0.70710678])
+    >>> qml.math.trace_distance(sigma, batch0)
+    array([0.5       , 0.        , 0.70710678])
+
+    .. seealso:: :func:`pennylane.qinfo.transforms.trace_distance`
+    """
+    # Cast as a c_dtype array
+    state0 = cast(state0, dtype=c_dtype)
+
+    # Cannot be cast_like if jit
+    if not is_abstract(state0):
+        state1 = cast_like(state1, state0)
+
+    if check_state:
+        _check_density_matrix(state0)
+        _check_density_matrix(state1)
+
+    if state0.shape[-1] != state1.shape[-1]:
+        raise qml.QuantumFunctionError("The two states must have the same number of wires.")
+
+    if len(state0.shape) == len(state1.shape) == 3:
+        if state0.shape[0] != 1 and state1.shape[0] != 1 and state0.shape[0] != state1.shape[0]:
+            raise ValueError(
+                "The two states must be batches of the same size, or one of them must contain a "
+                "single element."
+            )
+
+    eigvals, _ = qml.math.linalg.eigh(state0 - state1)

```

Comment:
Since we're throwing out the eigenvectors, we might as well just use `qml.math.linalg.eigvalsh` for better performance.

Reply:
This was also my first thought, but when trying to differentiate it when used with a `QNode`, I got a `NotImplementedError: VJP of eigvalsh wrt argnums (0,) not defined`

Reply:
That's strange 🤔. It's used in the implementation of `vn_entropy` and `fidelity`, and those all have correct gradients.

Reply:
Aha! It should be `qml.math.eigvalsh` instead of `qml.math.linalg.eigvalsh`. The former has a gradient defined but the latter doesn't .... your guess is as good as mine.